models/resnet/resnet_impl.hpp

```diff
@@ -0,0 +1,187 @@
+    inputChannel(0),
+    inputWidth(0),
+    inputHeight(0),
+    numClasses(0)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet(
+    const size_t inputChannel,
+    const size_t inputWidth,
+    const size_t inputHeight,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>(
+        std::tuple<size_t, size_t, size_t>(
+        inputChannel,
+        inputWidth,
+        inputHeight),
+        includeTop,
+        preTrained,
+        numClasses)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet(
+    std::tuple<size_t, size_t, size_t> inputShape,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    inputChannel(std::get<0>(inputShape)),
+    inputWidth(std::get<1>(inputShape)),
+    inputHeight(std::get<2>(inputShape)),
+    numClasses(numClasses)
+{
+  if (preTrained)
+  {
+    std::string home = getenv("HOME");
+    preTrainedPath = home + "/.cache/mlpack/models/weights/resnet/resnet" +
+        std::to_string(ResNetVersion) + ".bin";
+    if (Utils::PathExists(preTrainedPath, true) == false)
+    {
+      std::cout << "Downloading resnet" + std::to_string(ResNetVersion) +
+          ".bin to " + preTrainedPath << std::endl;
+      Utils::DownloadFile("resnet" + std::to_string(ResNetVersion) + ".bin",
+          preTrainedPath, "", false, false,
+          "http://models.mlpack.org/resnet/");
+    }
+
+    LoadModel(preTrainedPath);
+    return;
+  }
+
+  // Config for different Versions.
+  std::map<size_t, std::map<std::string, std::array<size_t, 4>>>::iterator
+      configFinder = ResNetConfig.find(ResNetVersion);
+  std::map<std::string, std::array<size_t, 4>> config = configFinder->second;
+  builderBlock = config.begin()->first;
+  numBlockArray = config.begin()->second;
+
+  if (configFinder == ResNetConfig.end())
+  {
+    mlpack::Log::Fatal << "Incorrect ResNet version. Possible Values are: 18, "
+        "34, 50, 101 and 152" << std::endl;
+  }
+
+  resNet.Add(new ann::Convolution<>(3, 64, 7, 7, 2, 2, 3, 3, inputWidth,
+      inputHeight));
+
+  mlpack::Log::Info << "Convolution: " << "(" << 3 << ", " <<
+      inputWidth << ", " << inputHeight << ")" << " ---> (";
+
+  // Updating input dimesntions.
+  inputWidth = ConvOutSize(inputWidth, 7, 2, 3);
+  inputHeight = ConvOutSize(inputHeight, 7, 2, 3);
+
+  mlpack::Log::Info << 64 << ", " << inputWidth << ", " << inputHeight
+      << ")" << std::endl;
+
+  resNet.Add(new ann::BatchNorm<>(64, 1e-5));
+  mlpack::Log::Info << "BatchNorm: " << "(" << 64 << ")" << " ---> ("
+        << 64 << ")" << std::endl;
+
```

Comment:
We can use Conv. Block Function for this.

Reply:
We would need to create another seq layer and add a identitiy layer at the start of network, should we do that ?

Reply:
Done